### ISSUE TITLE
using eigs with LinearMap gives wrong ground state.

### ISSUE THREAD

**Issue by shekhardevanshu** [Jutho/LinearMaps.jl#146]
I am trying to find the ground state energy using eigs method in julia.

```
m = 1 # how many states?

H = LinearMap(multiplyHPsi!, 2^N, issymmetric=true, ismutating=true)
@time u, v = eigs(H; nev=m, which=:SR) 
Energy = real(u[1]+ shiftE*N) # undoing the shift in energy
Psi = v
E = Energy

```
which gives

`E = 16.364294305817175`
I am expecting a similar value but with a negative sign.

the multiplyHPsi! function is this:

```
function multiplyHPsi!(nPsi::AbstractVector, Psi::AbstractVector) 
   
    length(nPsi)==length(Psi) || throw(DimensionMismatch())
    dimN = length(Psi)  # Dimension of the vector space
    N = convert(Int64,log2(dimN))  # Number of spins
    dimPsi = tuple(fill(2::Int,1,N)...) # dimsPsi =(2,2,2, ..., 2)
    a = collect(2:N)
    a = [a;1]             
    cyclperm = tuple(a...)  #tuple for cyclic permutation cyclperm = (2, 3, ... N, 1)
    Psi = reshape(Psi,dimPsi)
    nPsi = zeros(dimPsi)
    for n=1:N
        Psi = reshape(Psi,(4,2^(N-2))) 
        nPsi = reshape(nPsi,(4,2^(N-2)))
        nPsi += H2*Psi
        Psi = reshape(Psi,dimPsi)
        nPsi = reshape(nPsi,dimPsi)
        Psi = permutedims(Psi,cyclperm)
        nPsi = permutedims(nPsi,cyclperm)
    end
    nPsi = reshape(nPsi,dimN)
    return nPsi
end
```
I don't think there is a problem with this function as it works fine when predicting ground state energy with the power method. I am not sure what could be the problem while using eigs? In addition, for the first run, sometimes it sometimes gives NaN. I don't understand what am I missing?

**Comment by Jutho** [comment]
Your `multiplyHPsi!` is not really working in place. This line
```
    nPsi = zeros(dimPsi)
```
just allocates a new `nPsi`, thus forgetting the reference to the input argument, and returns that one instead. Hence, your first input argument, which should hold the result, is left unchanged. I think you want to use `fill!(nPsi, 0)` instead.

**Comment by shekhardevanshu** [comment]
Hi Jutho! Thank you for pointing that out. I made the changes, but although it doesn't give NaN anymore, it is giving 0 as the ground state energy for both `which=:SR` and `which=:LR`. I am not sure, but does `reshape` and `permutedims` functions also change the reference? If so, is there an alternative?

**Comment by Jutho** [comment]
Yes indeed, permutedims also changes the reference, it cannot act in place. If you implement it like this, why would you want an in-place / mutating method in the first place? Just keep your original implementation, remove the first argument, and set `ismutating=false`.

**Comment by shekhardevanshu** [comment]
Yes, it worked! Actually, this code is a part of a lecture given in 2016. It would have worked then and since I am new to Julia, I was having hard time updating it! But thanks for your help.

**Comment by Jutho** [comment]
If this ever worked, then it was a pure accident. This behaviour has not changed for as long as I am using Julia, which is since around 2012.

**Comment by shekhardevanshu** [comment]
I don't know how it worked then, but it does now!
